### PR TITLE
Add BasicSSHUserPrivateKey credential convertor

### PR DESCRIPTION
**Pull request by fydrah**
Hi,

I wrote a convertor for basic ssh user/private key. It supports ssh creds with or without passphrase.

I tested the plugin with Jenkins 2.121.1 (current docker lts image) on a Kubernetes 1.10.

Let me know if something needs to be improved

Regards

### CODE REVIEW

**Comment by Angelmmiguel** [comment]
Any news on this? I'm very interested on this feature.

**Comment by sbussetti** [comment]
bump

**Comment by agentgonzo** [comment]
Sorry, I've only just seen this. I'm away for a week but will have a look
week I get back unless @jnord wants to take a look in the meantime. I think
I had my GH notifications broken

On 17 Aug 2018 21:56, "Steve Bussetti" <notifications@github.com> wrote:

bump

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/jenkinsci/kubernetes-credentials-provider-plugin/pull/15#issuecomment-413986130>,
or mute the thread

File: pom.xml [hunk 1]
```diff
@@ -149,6 +154,12 @@
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>plain-credentials</artifactId>
       <optional>true</optional>
+    </dependency>
+      <!-- Basic Username PrivateKey -->
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>ssh-credentials</artifactId>
+      <version>1.13</version>
```
Comment (agentgonzo) [inline pom.xml]: `version` will be used from the `dependencyManagement` section, so isn't needed here

File: src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/convertors/BasicSSHUserPrivateKeyCredentialsConvertor.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        String username = SecretUtils.requireNonNull(SecretUtils.base64DecodeToString(usernameBase64), "basicSSHUserPrivateKey credential has an invalid username (must be base64 encoded UTF-8)");
+
+        String passphraseBase64 = null;
+        String passphrase = null; 
+        try {
+            passphraseBase64 = SecretUtils.getNonNullSecretData(secret, "passphrase", "basicSSHUserPrivateKey credential: failed to retrieve passphrase, assuming private key has an empty passphrase");
+        } catch(CredentialsConvertionException ex) {}
+
+        if (passphraseBase64 != null) {
+          passphrase = SecretUtils.requireNonNull(SecretUtils.base64DecodeToString(passphraseBase64), "basicSSHUserPrivateKey credential has an invalid passphrase (must be base64 encoded UTF-8)");
```
Comment (agentgonzo) [inline src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/convertors/BasicSSHUserPrivateKeyCredentialsConvertor.java]: indenting is off here

**Comment by agentgonzo** [comment]
Looks good - a couple of minor points. @jtnord - can you take a look too when you get a few minutes.

File: docs/examples/basic-ssh-username-private-key-passphrase.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
+apiVersion: v1
+kind: Secret
+metadata:
+# this is the jenkins id.
+  name: "jenkins-key"
+  labels:
+# so we know what type it is.
+    "jenkins.io/credentials-type": "basicSSHUserPrivateKey"
```
Comment (jglick) [inline docs/examples/basic-ssh-username-private-key-passphrase.yaml]: Suggest using `sshUserPrivateKey` to match the binding symbol in `withCredentials`.

File: docs/examples/basic-ssh-username-private-key-passphrase.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
...khncjJOTzJTaFBleFBlUmZGeHIwdWdYR1RRdnlUMEhuTQovUS8vVitMZHVQTVg4YjJBc096STByUWgrNGJqb2hPWnZLbUdLaXVQdjNlU3ZxcGkvcjYyMDhaVlRCampGdkJPClVRaE1iUFV5UjZ2TzFyeUZEd0JNd01xUTA2bGRrWEFyaEIrU0cwZFluT0tiLzZnMG5PMkJWRlVDZ1lFQXdCZUgKSEdOR3V4d3VtNjNVQWFxeVg2bFJTcEdHbTZYU0NCaHp2SFVQblZwaGdxN25uWk9HbDB6M1U0OWpyZUN2dXVFYwpmQTlZcXhKanpvWnk1ODcwS09YWTJrbHRscS9VLzRMcmIwazc1YWc2WlZiaTBvZW1BQ042S0NIdEUrWm0yZGFjCnJXOG9LV3BSVGJzdk1PWVV2U2pGMHU4QkNyZXN0cFJVRjk3N0tzOENnWUVBaWNiTEZDaks5K296cStlSktQRk8KZVo2QlU2WVdSMmplNVo1RDZpM0N5elQrM3doWHZFQ3pkNnlMcFhmckR5RWJQVEI1alVhY2JCMGxUbVdGYjNmYgpVSzZuODlia0NLTzJBYjkvWEtKeEFrUHpjZ0dtTUUrdkxSeDh3NXYyOVNUV0FXNzhyai9IOXltUGJxcVRhSjgyCkdRNStqQkkxU3c2R2VOQVcrOFAycExFQ2dZQXMvZFhCaW1jb3NDTWloNFplbFpLTjRXU082S0wwbGRRcDNVQk8KWmNTd2dGalNlUkQ2MFhEMnd5b3l3aVVBdHQyeUVjUFFNdS83c2FUNjNIYlJZS0hEYW9KdUxrQ2l5TEJFNEc4dwpjNkM1Mjd0QnZTWUhWWXBHQWdrOG1TV2tRWlRaZFBEaGxtVjd2ZEVwT2F5RjhYM3VDRHk5ZVFsdmJ6SGUyY01RCmpFT2I5UUtCZ0czalN4R2ZxTi9zRDhXOUJocFZyeWJDWGgyUnZoeE9CSkFGeDU4d1NXVGtSY1lTd3BkeXZtN3gKd2xNdGNFZFFnYVNCZXVCVTNIUFVkWUUwN2JRTkFsWU8wcDlNUW5zTEh6ZDJWOXlpQ1gxU3E1aUI2ZFFwSHh5aQpzRFpMWTJNeW0xblVKV2ZFNDdHQWN4Rlp0clZoOW9qS2NtZ2lIbzhxUFRrV2pGR1k3eGUvCi0tLS0tRU5EIFJTQSBQUklWQVRFIEtFWS0tLS0tCg=="
```
Comment (jglick) [inline docs/examples/basic-ssh-username-private-key-passphrase.yaml]: Yuck. [displaying this stuff](https://itnext.io/secrets-in-plain-text-13a98f54ef97)

File: src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/convertors/BasicSSHUserPrivateKeyCredentialsConvertor.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        String privateKeyBase64 = SecretUtils.getNonNullSecretData(secret, "privateKey", "basicSSHUserPrivateKey credential is missing the privateKey"); 
+        String privateKey = SecretUtils.requireNonNull(SecretUtils.base64DecodeToString(privateKeyBase64), "basicSSHUserPrivateKey credential has an invalid privateKey (must be base64 encoded UTF-8)");
+
+        String usernameBase64 = SecretUtils.getNonNullSecretData(secret, "username", "basicSSHUserPrivateKey credential is missing the username");
+        String username = SecretUtils.requireNonNull(SecretUtils.base64DecodeToString(usernameBase64), "basicSSHUserPrivateKey credential has an invalid username (must be base64 encoded UTF-8)");
+
+        String passphraseBase64 = null;
+        String passphrase = null; 
+        try {
+            passphraseBase64 = SecretUtils.getNonNullSecretData(secret, "passphrase", "basicSSHUserPrivateKey credential: failed to retrieve passphrase, assuming private key has an empty passphrase");
```
[the remark this answers is not in the document]
Comment (jglick) [reply src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/convertors/BasicSSHUserPrivateKeyCredentialsConvertor.java]: Indeed there is little reason to use a passphrase in this context at all. The sole purpose of a passphrase is to defend a private key against stolen drives, malicious scripts snooping around a compromised filesystem, etc. When the passphrase is part of the same _document_ (or K8S resource) as the secret key, obviously if you have one you have the other, so there is zero added security. The same is true of private keys stored in Jenkins using the traditional `DirectEntryPrivateKeySource`, of course.

We should support passphrases as a convenience for preëxisting keys which were created with one and maybe used in some laptop’s `~/.ssh/`, but document that new keys generated with the intent of being stored in K8S do not need one.

**Comment by fydrah** [comment]
Thanks for feedback, I'll try to push a first update before the end of the week.

> it is often better to create a JIRA in the Jenkins project and then link it to the PR

--> We could add a note about it ? (or maybe I missed it...)

File: pom.xml (hunk 1 above)
Comment (fydrah) [reply pom.xml]: so, should I remove this dependency completely ?

File: src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/convertors/BasicSSHUserPrivateKeyCredentialsConvertor.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import org.jenkinsci.plugins.variant.OptionalExtension;
+import com.cloudbees.jenkins.plugins.kubernetes_credentials_provider.CredentialsConvertionException;
+import com.cloudbees.jenkins.plugins.kubernetes_credentials_provider.SecretToCredentialConverter;
+import com.cloudbees.jenkins.plugins.kubernetes_credentials_provider.SecretUtils;
+import com.cloudbees.plugins.credentials.CredentialsScope;
+
+/**
+ * SecretToCredentialConvertor that converts {@link BasicSSHUserPrivateKey}.
+ */
+@OptionalExtension(requirePlugins={"ssh-credentials"})
```
[the remark this answers is not in the document]
Comment (fydrah) [reply src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/convertors/BasicSSHUserPrivateKeyCredentialsConvertor.java]: based on what you said previously, if we set `ssh-credentials` dependency optional into pom.xml, this will be used ?

File: docs/examples/basic-ssh-username-private-key-passphrase.yaml (hunk 3 above)
Comment (fydrah) [reply docs/examples/basic-ssh-username-private-key-passphrase.yaml]: I'm ok to rename it, but should we also rename every file/class/test for consistency ?

File: docs/examples/basic-ssh-username-private-key-passphrase.yaml (hunk 4 above)
Comment (fydrah) [reply docs/examples/basic-ssh-username-private-key-passphrase.yaml]: I thought that it should be a working example (so already base64 encoded).
If this is displayed this as a plain text priv key, users will have to encode key before using it.

File: src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/convertors/BasicSSHUserPrivateKeyCredentialsConvertor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (fydrah) [reply src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/convertors/BasicSSHUserPrivateKeyCredentialsConvertor.java]: I added a new method `getOptionalSecretData` based on your suggestions.

File: pom.xml (hunk 1 above)
Comment (jtnord) [reply pom.xml]: I do not care about `credentials-binding`, this plugin does not (should not at least) need nor should not depend on it. ... 
This is a *credential provider* and as such can be used for things that are not jobs.  I do not want this to become yet another kitchen sink plugin puling in all manner of dependencies that are not needed for the core use case.

File: src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/convertors/BasicSSHUserPrivateKeyCredentialsConvertor.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jtnord) [reply src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/convertors/BasicSSHUserPrivateKeyCredentialsConvertor.java]: yes, if you make the dependency on `ssh-credentials` optional then this would be the correct way to do it.

File: src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/SecretUtils.java [hunk 7]
```diff
@@ -171,7 +172,23 @@ public static String getNonNullSecretData(Secret s, String key, String exception
...
+     * @param s the Secret
+     * @param key the key to get the data for (which may be mapped to another key).
+     * @param exceptionMessage the detailMessage of the exception if the data for the key (or mapped key) was not
+     *            present.
+     * @return Optional data for specified key
+     * @throws CredentialsConvertionException if the data was not present.
+     */
+    public static Optional<String> getOptionalSecretData(Secret s, String key, String exceptionMessage) throws CredentialsConvertionException {
+        String mappedKey = getKeyName(s, key);
+        if (s.getData().containsKey(key) || s.getData().containsKey(mappedKey)) {
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/SecretUtils.java]: I _think_ you should just be able to simplify this whole function as

```
String data = s.getData().get(mappedKey))
return Optional.ofNullable(data)
```

File: docs/examples/basic-ssh-username-private-key-passphrase.yaml (hunk 3 above)
Comment (jglick) [reply docs/examples/basic-ssh-username-private-key-passphrase.yaml]: Note that the _existing_ types in this plugin do not match the symbols established in `credentials-binding`. I suppose if we wished to establish consistency we could continue to accept the current `jenkins.io/credentials-type`s but deprecate them and only document the new ones.

File: docs/examples/basic-ssh-username-private-key-passphrase.yaml (hunk 4 above)
Comment (jglick) [reply docs/examples/basic-ssh-username-private-key-passphrase.yaml]: Yes, I was just noting that you may want to link to the tool I mentioned as a way of helping people debug this mess.

File: docs/examples/basic-ssh-username-private-key-passphrase.yaml (hunk 3 above)
Comment (jtnord) [reply docs/examples/basic-ssh-username-private-key-passphrase.yaml]: @jglick an `@Symbol` referes to a conrete `binding` class.  That class can (and does) bind base classes that could be extended.  Hypothetically we could have an SSH key that has a URL instead of encoded key and the credentials-binding would still be able to bind it (and the credential implementation knows how to retreive that key from the URL), however we (this plugin) would not be able to create it as we need to bind to the specific implementation.  
Basically the Credential *implementation* class and the Credential *Binding* are 2 separate disjoint things.
